[RFC] prometheus: 2.30.3 -> 2.33.3

K900 opened the pull request:
###### Motivation for this change
New upstream release. Upstream switched to shipping the UI as an npm workspace, which node2nix does not support, so work around it by just installing everything, shoving it into a "build deps" derivation, then throwing it out. Yes, this is horrible. No, I have no better ideas.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

K900 committed:
prometheus: 2.30.3 -> 2.33.3

Now with extra crimes

SuperSandro2000 commented on pkgs/servers/monitoring/prometheus/default.nix:
```diff
@@ -1,97 +1,68 @@
-    nodejs = nodejs-14_x;
-    inherit (stdenv.hostPlatform) system;
-  };
+  webuiWorkspace = yarn2nix-moretea.mkYarnWorkspace {
```
I think that breaks eval.

K900 replied:
It does, but I'm not sure why... I'll look into it some more when I run out of other ideas on how to build this.

SuperSandro2000 replied:
You are using IFD (importing json from a derivation) which is forbidden.

K900 replied:
Oh. Yep, should have looked at what `mkYarnWorkspace` is doing internally. My assumption was it just needs the lockfile... Guess we have to vendor the package.json too if we go this route.

Ma27 commented:
@K900 do you still intend to work on this? %)

K900 commented:
I do want to get this done somehow, but I've been hoping to find a way to do it with less crimes. Any ideas?

Ma27 commented:
I don't see that much other options since both node2nix and npmlock2nix seem to be missing npm workspace support if we want a source-build of JS. But bth, if there's a usable prebuilt version of the frontend, we may want to check out how well it fits here.

K900 commented:
Yeah, that's the problem... It's compiled directly into the Go binary, and not as some strings in `.text`, but as _generated Go code_, that goes through the optimizer, and I think a good chunk of it is gzipped as well, so there's no way we're pulling it back out (and yes, I tried).

K900 commented:
So, updated.

- Prometheus 2.33.3, which is the latest upstream
- Should now (hopefully) not attempt to IFD
- More and more consistent comments to describe all the horrible fuckery required
- Added a patch to workaround upstream JS dependency breakage

K900 commented:
...aaand it seems impossible to make yarn2nix not IFD here, so here's a different horrible thing instead.

K900 commented:
We now have people from upstream around, so hopefully we won't have to do THIS.
